Spawner check for currently running controllers on shutdown

megantuttle opened the pull request:
This PR improves controller shutdown by getting the list of currently running controllers before attempting to stop and unload them.  Previously, shutdown simply stopped and unloaded the controllers that were spawned on launch. This behavior caused errors if a different controller was running at the time shutdown was requested.

Commit:
check for currently running controllers on shutdown

Commit:
cleaned up

bmagyar commented on controller_manager/src/controller_manager/controller_manager_interface.py:
```diff
@@ -108,13 +102,13 @@ def start_stop_controllers(names, st):
     resp = s.call(SwitchControllerRequest(start, stop, strictness))
     if resp.ok == 1:
         if st:
-            print("Started {} successfully".format(names))
```
I think your PR needs rebasing. Please don't change the prints.

megantuttle replied:
Why should the spawner not be responsible for shutting down all running controllers?  I don't see a problem with making that change, I'm just trying to understand it.

And I'm not sure how the prints ended up being changed, it must have been my text editor.  I'll fix those.

bmagyar reviewed (changes requested):
I see the general idea with the PR but it is a little off. The spawner should only be responsible for stopping/unloading the controllers it loaded/started.

You should take the intersection of currently running vs loaded/started controller and only act on those.

Commit:
fixed prints

mathias-luedtke reviewed (changes requested):
The `spawner` should not deal with any controller that it was not configured for.

> Previously, shutdown simply stopped and unloaded the controllers that were spawned on launch. This behavior caused errors if a different controller was running at the time shutdown was requested.

If this really causes errors, there is a bug that needs to be fixed.
The current PR would start controllers that have been stopped (or even unloaded) in between.

megantuttle commented:
Error output from the original issue I was trying to resolve:
`[ros_control_controller_spawner-2] killing on exit`
`[INFO] [1539901011.740469]: Shutting down spawner. Stopping and unloading controllers...`
`[INFO] [1539901011.741018]: Stopping all controllers...`
`[ERROR] [1539901011.743983745]: Could not stop controller 'pos_based_pos_traj_controller' since it is not running`
`[INFO] [1539901011.744493]: Unloading all loaded controllers...`
`[INFO] [1539901011.744950]: Trying to unload pos_based_pos_traj_controller`
`[INFO] [1539901011.782248]: Succeeded in unloading pos_based_pos_traj_controller`
`[INFO] [1539901011.782680]: Trying to unload force_torque_sensor_controller`
`[ERROR] [1539901011.785208211]: Could not unload controller with name force_torque_sensor_controller because it is still running`
`[INFO] [1539901011.785470]: Succeeded in unloading force_torque_sensor_controller`
`[INFO] [1539901011.785781]: Trying to unload joint_state_controller`
`[ERROR] [1539901011.788846453]: Could not unload controller with name joint_state_controller because it is still running`
`[INFO] [1539901011.789072]: Succeeded in unloading joint_state_controller`

I am running Kinetic on Ubuntu 16.04, and using the `ur_modern_driver` package to control a UR3.  I switch from the `pos_based_pos_traj_controller` (started by default) to the `vel_based_pos_traj_controller` using the `switch_controllers` service shortly after starting up the system by calling `ur3_ros_control.launch` and `ur3_moveit_planning_execution.launch`.

Why is `spawner` only intended to shut down the controllers it started?  Why shouldn't it stop any controllers that are running when shutdown is requested?

Commit:
check for intersection of running and loaded controllers

megantuttle committed:
changed list intersection method

mathias-luedtke commented on controller_manager/src/controller_manager/controller_manager_interface.py:
```diff
@@ -1,18 +1,15 @@
 #! /usr/bin/env python
```
please do not change this file

megantuttle replied:
Reverted in ee4cf92

mathias-luedtke commented on controller_manager/scripts/spawner:
```diff
@@ -41,6 +41,7 @@ import yaml
 from controller_manager_msgs.srv import *
 from std_msgs.msg import *
 import argparse
+import controller_manager.controller_manager_interface
```
please  use the same approach for listing controllers as used for the other services.

mathias-luedtke replied:
(or do not call the service at all, see https://github.com/ros-controls/ros_control/pull/364#issuecomment-431430987)

megantuttle replied:
Done. See ee4cf92

mathias-luedtke commented on controller_manager/scripts/spawner:
```diff
@@ -191,9 +200,6 @@ def main():
 
     rospy.loginfo("Controller Spawner: Loaded controllers: %s" % ', '.join(loaded))
 
-    if rospy.is_shutdown():
```
whiy did you remove this block?

megantuttle replied:
It seemed redundant since a shutdown hook is defined, and `rospy.spin()` simply keeps python from exiting until `rospy.core.is_shutdown()` is set (see [rospy/client.py](https://github.com/ros/ros_comm/blob/kinetic-devel/clients/rospy/src/rospy/client.py#L118)).

mathias-luedtke commented:
> Why is spawner only intended to shut down the controllers it started? Why shouldn't it stop any controllers that are running when shutdown is requested?

The intended use case for `spawner` is to provide node that manages a (sub)set of controllers.
So the life cycle of the controllers is bound to the lifecycle of the node.
For example, a user could prepare different launch file for different control set-ups.
`spawner` "owns" the controllers, so these should better not stopped manually.

Instead of adding the list call, it might be easier to just use `BEST_EFFORT` for the stopping call.

megantuttle committed:
switched service calls to follow convention, reverted interface to original

bmagyar commented:
Just to give some context, on a humanoid we'd usually have a separate set
of controllers for arms, legs/mobile base managed by different spawners but
within the same controller_manager as they belong to the same hw comm
system.

bmagyar commented:
@megantuttle will you have time to revamps this or would you prefer to close this PR? 
IMHO this PR still has the potential to make shutdown more robust to changes (if the spawned controller was stopped elsewhere).

megantuttle commented:
> @megantuttle will you have time to revamps this or would you prefer to close this PR?
> IMHO this PR still has the potential to make shutdown more robust to changes (if the spawned controller was stopped elsewhere).

@bmagyar I believe I've addressed all of the comments here, am I missing something?
